e2e/rdctl.e2e.spec.ts

```diff
@@ -255,6 +255,92 @@ test.describe('Command server', () => {
   });
 
   test.describe('rdctl', () => {
+    test.describe('config-file and parameters', () => {
+      test.describe("when the config-file doesn't exist", () => {
+        let parameters: string[];
+        const configFilePath = path.join(paths.appHome, 'rd-engine.json');
+        const backupPath = path.join(paths.appHome, 'rd-engine.json.bak');
+        const mvCommand = os.platform() === 'win32' ? 'ren' : 'mv';
+
+        test.beforeAll(async() => {
+          const dataRaw = await fs.promises.readFile(configFilePath, 'utf-8');
+
+          serverState = JSON.parse(dataRaw);
+          parameters = [`--password=${ serverState.password }`,
+            `--port=${ serverState.port }`,
+            `--user=${ serverState.user }`,
+          ];
+          try {
+            await spawnFile(mvCommand, [configFilePath, backupPath]);
+          } catch (err) {
+            console.log(`Error trying to ${ mvCommand } ${ configFilePath } ${ backupPath }: `, err);
+            expect(err).toBeUndefined();
+          }
+        });
+        test.afterAll(async() => {
+          try {
+            await spawnFile(mvCommand, [backupPath, configFilePath]);
+          } catch (err) {
+            console.log(`Error trying to ${ mvCommand } ${ backupPath } ${ configFilePath }: `, err);
+            expect(err).toBeUndefined();
+          }
+        });
+
+        test('it complains with no parameters,', async() => {
+          const { stdout, stderr, error } = await rdctl(['list-settings']);
+
+          expect(error).toBeDefined();
+          expect(stderr).toContain(`Error: open ${ configFilePath }: no such file or directory`);
+          expect(stderr).not.toContain('Usage:');
+          expect(stdout).toEqual('');
+        });
+
+        test('it works with all parameters,', async() => {
+          const { stdout, stderr, error } = await rdctl(parameters.concat(['list-settings']));
+
+          expect(error).toBeUndefined();
+          expect(stderr).toEqual('');
+          expect(stdout).toMatch(/"kubernetes":/);
+          const settings = JSON.parse(stdout);
+
+          expect(['version', 'kubernetes', 'portForwarding', 'images', 'telemetry', 'updater', 'debug', 'pathManagementStrategy']).toMatchObject(Object.keys(settings));
+        });
+        test("it complains when some parameters aren't specified", async() => {
+          for (let idx = 0; idx < parameters.length; idx += 1) {
```

Comment:
(This can now be expressed as `for (const idx in parameters)`, but I'm not sure it's clearer.)

Reply:
I tried that but typescript assumes `idx` is a string, and then it complains that `TS2345: Argument of type 'string' is not assignable to parameter of type 'number | undefined'.`.

If I try `const idx: number in parameters` I get `TS2404: The left-hand side of a 'for...in' statement cannot use a type annotation.`

And if I try `parameters.slice(0, idx as number)` I get `TS2352: Conversion of type 'string' to type 'number' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.`

There comes a time where it makes sense to stop fighting with TypeScript and go with the code I submitted.

Reply:
Argh. Yeah, not worth the trouble, then.

Reply:
@ericpromislow @mook-as The behavior you describe makes sense and I think you might want `for...of` instead of `for...in`:

> It is better to use a [for](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for) loop with a numeric index, [Array.prototype.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach), or the [for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop, because they will return the index as a number instead of a string, and also avoid non-index properties.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#difference_between_for...of_and_for...in